### DISCUSSION TITLE
Should the RCP designer have a project explorer (navigator) view?

### DISCUSSION THREAD

**Discussion by hvbtup** [eclipse-birt/birt#769] [Q&A] [answered]
I noticed (without this PR) that the RCP Designer does not contain the Navigator/Project Explorer. Is that intended?
I'm using <curent-project-folder>/res in the preferences for my ressources. 
How is the current project folder defined in the RCP Designer? This is hard to see without a navigator or project explorer view.

**Comment by wimjongman** [comment]
I think yes. This version of BIRT is purely the designer. You would open a local .rptdesign file and then it is opened in this version.

TBH I think it is more a test than anything else. Can you track down the origin of this perspective, through commit comments?

Maybe @qs-wang knows?

**Comment by qs-wang** [comment]
@wimjongman  you're right. @hvbtup  it was intended. 

There's 2 version of the Designer, the IDE/plugin version which is running in Eclispe IDE. This one has the Nav/Project explorer.
The RCP version is a pure eclipse RCP app which means there's some IDE features are missing, e.g. Nav/Project explorer. You need open the local file on your local disk.
Is Project explorer part of IDE still (the Nav was)? Or is it a standard alone plugin that we can integrate with RCP version? I need find time to recall it, and see if we can add a file explorer into RCP version now.

Seems possible by looking at this link?

https://www.eclipse.org/forums/index.php/t/1104846/

**Reply by wimjongman** [reply]
Project Explorer is on the same level as deprecated navigator

**Reply by qs-wang** [reply]
I think the major issue was that ResourceNavigator is in ui.ide plugin, so that if we include the ResourceNavigator in RCP, then it includes almost all IDE level features. This was the major reason that BIRT team created the RCP build at the first place. 

See project explorer may not be at the same level by looking at  this article at https://dagimnew.wordpress.com/2014/08/31/how-to-add-project-explorer-view-in-eclipse-rcp-application ? 

I will need find time to set up the development env,  and make a try.  Though generally speaking all applications are RCP, so the key point is that we need check how many extra dependencies will be bringing in if we include the explorer plugins.

**Reply by wimjongman** [reply]
You are right, they are probably not on the same level. Are you interested to return as a project committer, @qs-wang? We could certainly use someone with your experience in the project.

**Reply by qs-wang** [reply]
I am more than happy to be back. So glad to see the BIRT is still going after my first init comment into eclipse's cvs repo.

**Comment by hvbtup** [comment]
Do we know if anybody is actually *using* the RCP designer?

I discovered it only recently, and I like it - not for me personally, but for those of our customers who want to develop reports themselves. These people are ususally not "developers", but more or less "power users" with a basic SQL knowledge. For thesm, the "all-in-one" perspective is intimidating and the reduction to the essential in the RCP designer is good.

I don't think that the project explorer is strictly necessary. What *is* necessary is that one can determine and configure where the resources are searched.

**Reply by wimjongman** [reply]
I have no idea but if it is available then probably yes.

**Reply by wimjongman** [reply]
The location of the resources is described in the report design.

**Reply by hvbtup** [reply]
No, the rptdesign only contains the bare file names of the ressources - without a directory.

**Reply by UrsZeidler** [reply]
We use birt as report engine in https://github.com/SerNet/verinice 
and to provide our customers the possibility to edit or create reports we also use the birt.rcp.designer as base for our own report designer. This is needed because we use our own oda driver to connect to our middle-ware application server. 
We do not use the explorer view in our designer.

**Comment by qs-wang** [comment]
@hvbtup I made a quick try, and seems it's possible to have project explorer with RCP. Though we have to manage the to include the minimal plugins to make the project explorer works all right and won't bring in to many other features. 

See attached the screen, the project explorer is included in the RCP application, however the view is not working 100% propert
<img width="1577" alt="Screen Shot 2021-12-12 at 4 08 48 pm" src="https://user-images.githubusercontent.com/32129299/145705300-f1cc8c00-c0d5-4676-bd62-523308fa4114.png">
y

**Comment by ruspl-afed** [comment]
The key difference is between CNF framework itself ( located at "org.eclipse.ui.navigator")
and its implementation for IWorkspace (located at "org.eclipse.ui.navigator.resources")

CNF should not add any IDE dependency, just regular E3 Workbench.
While `org.eclipse.ui.navigator.ProjectExplorer` view (implemented by `org.eclipse.ui.navigator.resources.ProjectExplorer` class) will surely bring a lot of IDE-related contributions.

**Reply by qs-wang** [reply]
Making sense. Any idea of what's the minimal dependencies? Maybe make a quick try to see what's the minimal IDE-related features will be added?

**Reply by ruspl-afed** [reply]
I wonder what do we really want to "explore" with this explorer - this is the first question that we need to answer. 
After that we can provide our own explorer to show what is needed.

For initial question I would answer "No", because "Project Explorer" wants IDE projects to "explore" them, that is not the case for RCP designer that works with standalone user files.

TBH, I'm saying this without deep digging into the BIRT sources regarding the matter of discussion

**Comment by wimjongman** [comment, marked as answer]
The RCP designer should not have an explorer since it is intended to open .rptdesign files from the file system.

Please note that there are more BIRT RCP instances. There is the all-in-one which includes an explorer/

**Reply by hvbtup** [reply]
This still leaves the question open: How can the user determine and configure where the resources are searched? Note that it would be OK for me even if it cannot be configured - but then it should be documented at least.

**Reply by ruspl-afed** [reply]
> How can the user determine and configure where the resources are searched

@hvbtup Please post a screenshot with example of what we are talking about.

**Comment by hvbtup** [comment]
This is so embarassing...

In the "All-In-One" designer, you can specify the resource path in the menu under Window / Preferences / Report Design / Resource.

I was going to write: """ In the "RCP Designer" there is no such menu" """
and just wanted to confirm this statement.
I could *swear* the menu was missing when I created the discussion. 

But now it's there:
![grafik](https://user-images.githubusercontent.com/7644776/147932408-e0844272-5081-40e3-a4ab-d1e0207f0c28.png)

Probably this is a positive consequence of my commits regarding https://github.com/eclipse/birt/issues/694?

Anyway, as long as the menu is available, that's fine for me and I consider this solved.
